### DISCUSSION TITLE
Unregistered an array index keeps values as 'undefined'

### DISCUSSION THREAD

**Discussion by asaf** [react-hook-form/react-hook-form#4907] [Q&A] [answered]
Hey,

Trying to remove the first element yields: `[undefined,"bar"]`
(can't use a hook such as useFieldArray as I have complex dynamic form)

https://codesandbox.io/s/react-hook-form-v7-form-context-forked-tps7y?file=/src/index.js

(In the dom the second element is kept as `list-1`).

What would be the correct way to remove an item from an array?

Thanks!

**Comment by bluebill1049** [comment]
don't do it yourself, it probably going to be quite painful. use `useFieldArray`, it have sort out most of the edge cases as well.

**Reply by miggynascimento** [reply]
Hey @asaf, I struggled with this personally until I got useFieldArray to work. @bluebill1049 is correct that it is painful to try to implement array yourself.

I modified your codesandbox to show how you may be able to achieve what you need using `useFieldArray`
You won't need `watch`, `setValue`, nor `unregister` with useFormArray.
All you need to do is pass the `control` or if it's deeply nested you can use the context to grab the `control`.

https://codesandbox.io/s/react-hook-form-v7-form-context-forked-2bot2?file=/src/index.js

Hope this helps, cheers

**Reply by bluebill1049** [reply]
Thank you @miggynascimento i would have voted yours as an answer :) (if you push to a separate comment) 💪

**Reply by miggynascimento** [reply]
ah oops, first time answering. thanks for the great library!

**Comment by miggynascimento** [comment, marked as answer]
I modified your codesandbox to show how you may be able to achieve what you need using useFieldArray
You won't need watch, setValue, nor unregister with useFormArray.
All you need to do is pass the control or if it's deeply nested you can use the context to grab the control.

https://codesandbox.io/s/react-hook-form-v7-form-context-forked-2bot2?file=/src/index.js

**Comment by asaf** [comment]
Thanks guys, as i mentioned I can't use useFieldArray, my renderers happen conditionally and I don't know the field names ahead as they are generated automatically from a schema.
